Add missing net_buf_unref for non-recoverable errors while sending l2cap data

prathje opened the pull request:
In case the connection breaks while transmitting, the net_buf while be dequeued but not unreferenced.

jhedberg reviewed (changes requested):
Looks fine, but you’ll need to have a proper prefix to the commit message summary (CI will probably also fail because of it). Something like “Bluetooth: L2CAP: Fix missing net_buf_unref()” should be sufficient, with more details in the commit message body.

prathje commented:
I updated the commit message and also removed the trailing whitespace to hopefully get through compliance.

prathje committed:
Bluetooth: L2CAP: Fix missing net_buf_unref()

In case of non-recoverable errors (e.g. the connection breaks while
transmitting), the l2cap_chan_tx_process deques the net_buf but does
not unreference it. As this is inside a work queue thread, the sending
thread gets no information on this error, relying on the tx_process to
ultimately free the buffer.

Signed-off-by: Patrick Rathje <git@patrickrathje.de>

prathje commented:
Sign-off was missing for compliance.

jhedberg commented:
Does `bt_l2cap_chan_send()` require a similar fix (end of the function)?

jhedberg commented:
@Vudentz @joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago 😄

Thalley commented:
> @Vudentz @joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago

@jhedberg Perhaps we should add the DNM label if we want their input first

jhedberg commented:
> @jhedberg Perhaps we should add the DNM label if we want their input first

@Thalley I dismissed my own review until then, which does the same trick in practice (can't merge now)

joerchan commented:
> Vudentz joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago

So this is a regression due to that change. I suggest to add this comment above the net_buf_unref .
```
				/* The ownership of the buffer is with the stack
				 * since bt_l2cap_chan_send has returned success
				 */
```

prathje commented:
@joerchan I think a comment won't hurt, so I just went ahead and added it 👍

joerchan commented:
@prathje Thanks. I think you could squash the two commits though.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -872,6 +872,11 @@ static void l2cap_chan_tx_process(struct k_work *work)
 		if (sent < 0) {
 			if (sent == -EAGAIN) {
 				ch->tx_buf = buf;
+			} else {
+				/* The ownership of the buffer is with the stack
+				 * since bt_l2cap_chan_send has returned success
+				 */
+				net_buf_unref(buf);
```
Sorry, but this is a bit confusing to me. On two levels.

Firstly, these branches handle the return value of l2cap_chan_le_send_sdu(), so why the reference to bt_l2cap_chan_send() in the comment? Was that a typo or intentional?

Secondly, "since bt_l2cap_chan_send has returned success" looks very odd considering that we're in a branch where l2cap_chan_le_send_sdu() has *not* returned success. I'd have worded it the other way around "If l2cap_chan_le_send_sdu() fails the ownership of the buffer is retained by the caller, so we must unref it" or something like that.

prathje replied:
Yes, this is intentional. The ownership has been transferred by a previous user call of `bt_l2cap_chan_send` (not the internal `l2cap_chan_le_send_sdu`). `bt_l2cap_chan_send` returned success to the caller, otherwise the buffer would not have been queued.

prathje replied:
Comment was removed again. So I resolve this :)

prathje commented:
> @prathje Thanks. I think you could squash the two commits though.

Yes, I have thought of that but decided not to as the first commit has been merged already (https://github.com/nrfconnect/sdk-zephyr/pull/555). I am happy to merge if you still prefer it?

carlescufi commented:
> > @prathje Thanks. I think you could squash the two commits though.
> 
> Yes, I have thought of that but decided not to as the first commit has been merged already ([nrfconnect/sdk-zephyr#555](https://github.com/nrfconnect/sdk-zephyr/pull/555)). I am happy to merge if you still prefer it?

That has not been merged to Zephyr, this commit was cherry-picked into a fork of the Zephyr repo.

joerchan commented:
@prathje That's not a problem, what Carles said is correct, so please squash them. But thank you for being considerate.
If you keep the first commit message it will go smoothly.

carlescufi commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -872,6 +872,11 @@ static void l2cap_chan_tx_process(struct k_work *work)
 			if (sent == -EAGAIN) {
 				ch->tx_buf = buf;
+			} else {
+				/* The ownership of the buffer is with the stack
+				 * since bt_l2cap_chan_send has returned success
```
Please remove the comment, since this is self-explanatory by the fact that this is an internal processing function that takes net_buf references from an internal queue, and therefore those are owned by the stack and not the application.

joerchan replied:
Ok, so the comment generated more confusion than clarity, so we discussed it and decided it was best to just remove it.

prathje replied:
Will do :)

prathje commented:
@carlescufi @joerchan Thanks! With the requested comment removal, we are back to just the first commit anyway. I also did a quick rebase 👍
